homeassistant/components/rainbird/switch.py

```diff
@@ -4,58 +4,83 @@
+from . import DATA_RAINBIRD, DOMAIN, RAINBIRD_CONTROLLER
 
-from . import DATA_RAINBIRD
-
-DOMAIN = "rainbird"
 _LOGGER = logging.getLogger(__name__)
 
-PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+ATTR_DURATION = "duration"
+
+SERVICE_START_IRRIGATION = "start_irrigation"
+
+SERVICE_SCHEMA_IRRIGATION = vol.Schema(
     {
-        vol.Required(CONF_SWITCHES, default={}): vol.Schema(
-            {
-                cv.string: {
-                    vol.Optional(CONF_FRIENDLY_NAME): cv.string,
-                    vol.Required(CONF_ZONE): cv.string,
-                    vol.Required(CONF_TRIGGER_TIME): cv.string,
-                    vol.Optional(CONF_SCAN_INTERVAL): cv.string,
-                }
-            }
-        )
+        vol.Required(ATTR_ENTITY_ID): cv.entity_id,
+        vol.Required(ATTR_DURATION): vol.All(vol.Coerce(float), vol.Range(min=0)),
     }
 )
 
 
 def setup_platform(hass, config, add_entities, discovery_info=None):
     """Set up Rain Bird switches over a Rain Bird controller."""
-    controller = hass.data[DATA_RAINBIRD]
 
+    if discovery_info is None:
+        return False
+
+    controller: RainbirdController = hass.data[DATA_RAINBIRD][
+        discovery_info[RAINBIRD_CONTROLLER]
+    ]
+    available_stations: AvailableStations = controller.get_available_stations()
+    if not (available_stations and available_stations.stations):
+        return False
     devices = []
-    for dev_id, switch in config.get(CONF_SWITCHES).items():
-        devices.append(RainBirdSwitch(controller, switch, dev_id))
+    for i in range(1, available_stations.stations.count + 1):
+        if available_stations.stations.active(i):
+            time = discovery_info.get("zones", {}).get(
+                CONF_TRIGGER_TIME, discovery_info.get(CONF_TRIGGER_TIME, 0)
+            )
+            name = discovery_info.get("zones", {}).get(CONF_FRIENDLY_NAME)
+            if time:
+                devices.append(RainBirdSwitch(controller, i, time, name=name))
+            else:
+                logging.warning(
+                    "No delay configured for zone {0:d}, controller {1:s}. "
+                    "Not adding sprinklers for zone {0:d}.".format(
+                        i, discovery_info[RAINBIRD_CONTROLLER]
+                    )
+                )
+
     add_entities(devices, True)
 
+    def _start_irrigation(service):
+        entity_id = service.data.get(ATTR_ENTITY_ID)
+        duration = service.data.get(ATTR_DURATION)
+
+        for d in devices:
+            if d.entity_id == entity_id:
+                d.turn_on(duration=duration)
+
+    hass.services.register(
+        DOMAIN,
+        SERVICE_START_IRRIGATION,
+        _start_irrigation,
+        schema=SERVICE_SCHEMA_IRRIGATION,
+    )
+
 
 class RainBirdSwitch(SwitchDevice):
     """Representation of a Rain Bird switch."""
 
-    def __init__(self, rb, dev, dev_id):
+    def __init__(self, controller: RainbirdController, zone, time, name=None):
```

Comment:
We don't need to make name optional.
```suggestion
    def __init__(self, controller: RainbirdController, zone, time, name):
```

Reply:
done